src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private Object[] array;
+    private int size;
+
+    public ArrayList() {
+        this.array = new Object[DEFAULT_CAPACITY];
+        this.size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == array.length) {
+            newSizeForArray();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + " , Size: " + size);
+        }
+        if (size == array.length) {
+            newSizeForArray();
+        }
+        for (int i = size; i > index; i--) {
+            array[i] = array[i - 1];
+        }
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() > array.length) {
+            newSizeForArray();
+        }
+        for (int i = 0; i < list.size();i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + " , Size: " + size);
+        }
+        return (T) array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + " , Size: " + size);
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + " , Size: " + size);
+        }
+        final T removedElement = (T) array[index];
+        for (int i = index; i < size - 1; i++) {
+            array[i] = array[i + 1];
+        }
+        array[size - 1] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(array[i], element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException();
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size <= 0;
```

Comment:
Probably, the size will always be >= 0;
```suggestion
        return size == 0;
```